test/StripeResource.spec.js

```diff
@@ -32,4 +35,245 @@ describe('StripeResource', function() {
+      },
+      params: 'amount=1000&currency=usd&source=tok_visa&description=test'
+    };
+
+    afterEach(function() {
+      realStripe.setMaxNetworkRetries(0);
+      stripe.setMaxNetworkRetries(0);
+    });
+
+    after(function() {
+      nock.cleanAll();
+    })
+
+    describe('_request', function() {
+      it('throws an error on connection failure', function(done) {
+        // Mock the connection error.
+        nock('https://' + options.host)
+          .post(options.path, options.params)
+          .replyWithError('bad stuff');
+
+        realStripe.charges.create(options.data, function(err) {
+          expect(err.detail.message).to.deep.equal('bad stuff');
+          done();
+        });
+      });
+
+      it('should retry the connection if max retries are set', function(done) {
+        nock('https://' + options.host)
+          .post(options.path, options.params)
+          .replyWithError('bad stuff')
+          .post(options.path, options.params)
+          .replyWithError('worse stuff');
+
+        realStripe.setMaxNetworkRetries(1);
+
+        realStripe.charges.create(options.data, function(err) {
+          var errorMessage = realStripe.invoices._generateConnectionErrorMessage(1);
+          expect(err.message).to.equal(errorMessage);
+          done();
+        });
+      });
+
+      it('should stop retrying after a successful retry', function(done) {
+        nock('https://' + options.host)
+          .post(options.path, options.params)
+          .replyWithError('bad stuff')
+          .post(options.path, options.params)
+          .reply(200, {
+            id: 'ch_123',
+            object: 'charge',
+            amount: 1000,
+          });
+
+        realStripe.setMaxNetworkRetries(2);
+
+        realStripe.charges.create(options.data, function(err, charge) {
+          expect(charge.id).to.equal('ch_123');
+          done();
+        });
+      });
+
+      it('should retry on rate limit error', function(done) {
+        nock('https://' + options.host)
+          .post(options.path, options.params)
+          .reply(429, {
+            error: {
+              message: 'Rate limited'
+            }
+          })
+          .post(options.path, options.params)
+          .reply(200, {
+            id: 'ch_123',
+            object: 'charge',
+            amount: 1000,
+          });
+
+        realStripe.setMaxNetworkRetries(1);
+
+        realStripe.charges.create(options.data, function(err, charge) {
+          expect(charge.id).to.equal('ch_123');
+          done();
+        });
+      });
+
+      it('should not retry on a 400 error', function(done) {
+        nock('https://' + options.host)
+          .post(options.path, options.params)
+          .reply(400, {
+            error: {
+              message: 'You goofed'
+            }
+          });
+
+        realStripe.setMaxNetworkRetries(1);
+
+        realStripe.charges.create(options.data, function(err) {
+          expect(err).to.not.be.null;
```

Comment:
Hmm... I think this test could pass erroneously (in theory). Can we explicitly check that it generates the expected error message, instead of some other error? Eg; check that `you goofed` appears in the right place. 

Same with the 500 test below, so we know we're passing along the right message, not blowing up internally